From c32e5e17a0530f04c5be96cee30684f869f13d35 Mon Sep 17 00:00:00 2001
From: Gaosheng Cui <cuigaosheng1@huawei.com>
Date: Tue, 18 Apr 2023 08:49:14 -0400
Subject: [PATCH 0085/1646] [Huawei] Revert "[Backport] crypto: api - Do not
 create test larvals if manager is disabled"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188696

--------------------------------

This reverts commit 1050a6bf5a177c1739882c6dc1bba2e1eee3abcb.

The patch set of 1050a6bf5a17 ("crypto: api - Do not create test
larvals if manager is disabled") conflicts with the SIMD feature,
and maybe cause kernel panic on arm64, so revert the patch set.

Fixes: 1050a6bf5a17 (crypto: api - Do not create test larvals if manager is disabled)
Signed-off-by: Gaosheng Cui <cuigaosheng1@huawei.com>
---
 crypto/algapi.c | 56 ++++++++++++++++++-------------------------------
 crypto/api.c    |  7 +++++--
 2 files changed, 25 insertions(+), 38 deletions(-)

diff --git a/crypto/algapi.c b/crypto/algapi.c
index f6481cb79946..d6cd860bca4f 100644
--- a/crypto/algapi.c
+++ b/crypto/algapi.c
@@ -216,32 +216,6 @@ void crypto_remove_spawns(struct crypto_alg *alg, struct list_head *list,
 }
 EXPORT_SYMBOL_GPL(crypto_remove_spawns);
 
-static struct crypto_larval *crypto_alloc_test_larval(struct crypto_alg *alg)
-{
-	struct crypto_larval *larval;
-
-	if (!IS_ENABLED(CONFIG_CRYPTO_MANAGER))
-		return NULL;
-
-	larval = crypto_larval_alloc(alg->cra_name,
-				     alg->cra_flags | CRYPTO_ALG_TESTED, 0);
-	if (IS_ERR(larval))
-		return larval;
-
-	larval->adult = crypto_mod_get(alg);
-	if (!larval->adult) {
-		kfree(larval);
-		return ERR_PTR(-ENOENT);
-	}
-
-	refcount_set(&larval->alg.cra_refcnt, 1);
-	memcpy(larval->alg.cra_driver_name, alg->cra_driver_name,
-	       CRYPTO_MAX_ALG_NAME);
-	larval->alg.cra_priority = alg->cra_priority;
-
-	return larval;
-}
-
 static struct crypto_larval *__crypto_register_alg(struct crypto_alg *alg)
 {
 	struct crypto_alg *q;
@@ -276,20 +250,31 @@ static struct crypto_larval *__crypto_register_alg(struct crypto_alg *alg)
 			goto err;
 	}
 
-	larval = crypto_alloc_test_larval(alg);
+	larval = crypto_larval_alloc(alg->cra_name,
+				     alg->cra_flags | CRYPTO_ALG_TESTED, 0);
 	if (IS_ERR(larval))
 		goto out;
 
-	list_add(&alg->cra_list, &crypto_alg_list);
+	ret = -ENOENT;
+	larval->adult = crypto_mod_get(alg);
+	if (!larval->adult)
+		goto free_larval;
 
-	if (larval)
-		list_add(&larval->alg.cra_list, &crypto_alg_list);
+	refcount_set(&larval->alg.cra_refcnt, 1);
+	memcpy(larval->alg.cra_driver_name, alg->cra_driver_name,
+	       CRYPTO_MAX_ALG_NAME);
+	larval->alg.cra_priority = alg->cra_priority;
+
+	list_add(&alg->cra_list, &crypto_alg_list);
+	list_add(&larval->alg.cra_list, &crypto_alg_list);
 
 	crypto_stats_init(alg);
 
 out:
 	return larval;
 
+free_larval:
+	kfree(larval);
 err:
 	larval = ERR_PTR(ret);
 	goto out;
@@ -418,11 +403,10 @@ int crypto_register_alg(struct crypto_alg *alg)
 	down_write(&crypto_alg_sem);
 	larval = __crypto_register_alg(alg);
 	test_started = static_key_enabled(&crypto_boot_test_finished);
-	if (!IS_ERR_OR_NULL(larval))
-		larval->test_started = test_started;
+	larval->test_started = test_started;
 	up_write(&crypto_alg_sem);
 
-	if (IS_ERR_OR_NULL(larval))
+	if (IS_ERR(larval))
 		return PTR_ERR(larval);
 
 	if (test_started)
@@ -632,8 +616,8 @@ int crypto_register_instance(struct crypto_template *tmpl,
 	larval = __crypto_register_alg(&inst->alg);
 	if (IS_ERR(larval))
 		goto unlock;
-	else if (larval)
-		larval->test_started = true;
+
+	larval->test_started = true;
 
 	hlist_add_head(&inst->list, &tmpl->instances);
 	inst->tmpl = tmpl;
@@ -642,7 +626,7 @@ int crypto_register_instance(struct crypto_template *tmpl,
 	up_write(&crypto_alg_sem);
 
 	err = PTR_ERR(larval);
-	if (IS_ERR_OR_NULL(larval))
+	if (IS_ERR(larval))
 		goto err;
 
 	crypto_wait_for_test(larval);
diff --git a/crypto/api.c b/crypto/api.c
index 0e7a255252ca..44d3e4f8bb58 100644
--- a/crypto/api.c
+++ b/crypto/api.c
@@ -166,8 +166,11 @@ void crypto_wait_for_test(struct crypto_larval *larval)
 	int err;
 
 	err = crypto_probing_notify(CRYPTO_MSG_ALG_REGISTER, larval->adult);
-	if (WARN_ON_ONCE(err != NOTIFY_STOP))
-		goto out;
+	if (err != NOTIFY_STOP) {
+		if (WARN_ON(err != NOTIFY_DONE))
+			goto out;
+		crypto_alg_tested(larval->alg.cra_driver_name, 0);
+	}
 
 	err = wait_for_completion_killable(&larval->completion);
 	WARN_ON(err);
-- 
2.17.1

